#691 Updates to config functionality to make it easier to run great3-like sims

rmjarvis opened the pull request:
This is a pretty big pull request.  Not only have I added a bunch of new features to the config apparatus to make it more functional, but I also took the opportunity to completely redesign the back end config layer.  In fact, since the PR was originally way too big for GitHub's diff screen to show all of it, I put a lot of the boring functionality-neutral changes on master already so the diff here mostly shows the interesting new features that are worth code reviewing.  But for full disclosure, here is a [listing of the changes that are already on master](https://github.com/GalSim-developers/GalSim/compare/7bd70046a731588181c3e25da32b401cac4d5a72...e0cd6920b75bd867ce6175cf96d698b3f6459486).

Updates to galsim executable:
- Dropped default verbosity from 2 to 1, since for real simulations, 2 is usually too much output.
- Added ability to easily split the total work into several jobs with `galsim -n njobs -j jobnum`.  This will just do a single job numbered jobnum out of a total of njobs jobs.  This makes it much easier to split the work of a single config file over several machines.
- Added `galsim -p` to perform profiling on the run.  Sometimes that's helpful to see what might be taking most of the running time.  Although, I've found that the output isn't always as enlightening as I would have hoped.

New config features:
- Truth catalogs can be output as specified in an `output.truth` field.  cf. demo9.yaml, demo10.yaml, cgc.yaml
- External modules can be specified in the config file in a top-level 'modules' field, rather than just with `galsim -m module`.  cf. examples/des/draw_psf.yaml
- Fields can be specified using `template` to read either a whole file or just a given field from a file and use that for the field in question.  This can significantly reduce repeated code for similar simulations.  cf. rgc.yaml, cgc_psf.yaml
- If a value field for something is a string and it starts with "$", then the rest of the string is treated as an Eval string.  Also, the character "@" in an Eval string is the equivalent of a Current type specification.  This ends up being very readable in most cases.  Often better than explicitly using the Eval and Current types for everything.  cf. cgc.yaml 
- In addition to values that are generated elsewhere in the config, now gsobjects can also be referenced from an Eval string, either using the @ notation or a regular Current type.  cf. demo10.yaml, output.truth.columns.cosmos_index.xgal
- A RandomDistribution can now specify a LookupTable using x and f lists, rather than requiring either a file with these columns or a lambda function.  cf. cgc.yaml, psf.items.0.fwhm

Notable features of the backend redesign:
- The config file is now read as an OrderedDict.  This shouldn't matter for anything except the truth catalog, which can now have the output columns in the same order as specified in the config file.
- I made pretty much everything customizable.  There is a basic concept of input objects, output files, images, stamps, gsobjects, values, etc, but the exact implementation of each is something the user can change if they want.  They can do this by writing their own "type" for anything they want to work differently and then register it with the appropriate Register function.  e.g. RegisterImageType will let you define a new `image.type` and specify how it works in each of several point in the code (setup, build, adding noise, number of objects).  (Note, this had been partially true previously, but the redesign makes it much easier and extends the modularity to more things.)  I think this enables all of the custom python code extensions envisioned by @TallJimbo for Great4 sims.
- There had been a lot of repeated code, which I've now made into helper functions that are called by multiple functions.  So most of the code is now packaged in smaller, more focused functions, which are easier to understand, I hope.  Also, for people designing their own type, they can probably use some of these functions as well.  (TODO: A task for me here is to write some good documentation on the wiki explaining how one might do this.  E.g. @esheldon already asked if it would be possible to draw from a log-normal distribution, so I explained how he could write his own value generator to do this.  Better if I could just point to good documentation about this.)
- All the logger commands are now guarded to check whether they will produce output.  This looks redundant, but when the loggers are communicating through a proxy, the arguments had to be sent through the multiprocessing pipe to get to the real logger object, who would usually then not do anything with it, which is inefficient.  So now I guard them all with an appropriate if statement to see if they will produce any output, which sped things up significantly in some cases.
- I changed how config handles random number seeds.  It had been the case that the objects in a file would use (seed + 0 .. seed + nobj-1), and anything at the file or image level would use (seed + nobj).  However, this wasn't really a good solution, since you might need the rng to determine what nobj is.  In fact, we do this in demo9.  So the rng seeding was a bit incoherent in this case.  It seemed better to change to using (seed + 0) for all file/image-level needs, and use (seed + 1 ... seed + nobj) for the objects.
- I don't pass the weight, badpix, and psf images around anymore (which are usually None) as return values from BuildImages and similar commands.  Now these are handled using a new "extra_output" concept.  There are now hooks into various places in the code where these "extra" output types can do some additional processing.  The images they make if they are active are stored in a special spot in the config dict, so final processing and output can be done at the top level.  This means that one can register additional extra output types that do other things than these.  See extra*.py for more details.
- I also added the concept of a stamp type.  Currently, there is only one stamp type, which I called Basic.  But there are now hooks into the stamp creation to do things differently.  e.g. @beckermr wants to be able to draw blended objects on stamps, which hadn't been possible, but now we can make a custom stamp type that would do this.  TODO: I've also promised him an example that will showcase this feature so he can see how to extend it to other similar simulations.

Updates to demos:
- Added examples/great3 directory with cgc.yaml, cgc_psf.yaml, rgc.yaml, rgc_psf.yaml.  I was originally planning on doing all 20, including the deep fields, but this PR is already too big, so I think I'll make another issue to flesh out the rest of these.
- Added truth catalog output in demo9, demo10.
- Introduce Current type in demo10.
- Moved explanation of `eval_variables` from demo10 to demo11.
- Set image.nproc : -1 in demo11, which didn't used to work correctly for subtle reasons, that I fixed on this branch.  (Related to the rng syncing for PowerSpectrum.buildGrid.  See below.)
- Updated demo8.py to reflect the simpler usage that is now possible with the refactored config code.
- Updated all the demo*.py files to match the new random number seed convention.
- Removed most of the comment lines from json files.  cf. discussion here: #301 
- A few minor changes to the order of some calculations in the .py files to match up with some refactoring changes I made in the config layer.

Updates to regular GalSim code:
- Added OutputCatalog class, used to implement truth catalogs.
- galsim.fits.writeMulti allows some "images" to be already-built hdus that are included as is.  Used for the truth catalog hdu in demo10.
- image.resize now allows an optional wcs argument
- Made a method equivalent for image[bounds] = rhs, called image.setSubImage(bounds,rhs).  Useful if accessing an image through a proxy object that doesn't allow the regular syntax.
- Added a function PowerSpectrum.nRandCallsForBuildGrid, which returns how many calls to the rng were made by ps.buildGrid.  Used to keep the rng synced when the ps is a proxy object, so the rng used by buildGrid is not the same as the local rng.
- Added rng.discard and rng.raw.  The former allows one to quickly discard a large number of random values from the rng.  Used in conjunction with ps.nRandCallsForBuildGrid() to keep the rng in sync when multiprocessing. For reference, discarding 1 million values takes 3 ms on my laptop.  So it's pretty quick.  rng.raw() isn't really used here, but it was a feature I wanted recently in another context, so I added it too. It returns the raw long int value generated by the underlying random number generator, not converted to a float.
- Added min/max hlr and min/max flux for COSMOSCatalog.  Currently this gives a rough approximation to the S/N cuts made by Great3 in cgc.yaml.  But it really should be improved further to allow for cuts that are closer to what Great3 did.  cf. Issue #693.
- Added a `sersic_prec` option to COSMOSCatalog.makeGalaxy to round off the Sersic index n values to some precision.  Since GalSim caches a lot of information about each Sersic n it uses, using fewer n values runs a lot faster.  Default is 0.05, which I imagine is usually plenty accurate.

rmandelb commented:
Whoa.  OK, I will try to take a look at this over break. I assume this is the branch you were recommending I work off of when I do #693, but please let me know if I've misunderstood.

rmjarvis commented:
>  I assume this is the branch you were recommending I work off of when I do #693, but please let me know if I've misunderstood.

Yes.  Ideally, on that branch we can try to make cgc.yaml and rgc.yaml really produce something equivalent to the Great3 branches.  I _think_ the only thing missing is the selection of galaxies from the COSMOS catalog, which so far is only an approximation to what was done for Great3.

rmandelb commented:
Should I be getting a "WARNING: Overwriting existing file 'output/catalog.fits'. [astropy.io.fits.file]" when running the unit tests?

rmjarvis commented:
> Should I be getting a "WARNING: Overwriting existing file 'output/catalog.fits'. [astropy.io.fits.file]" when running the unit tests?

I don't get that.  But maybe depends on the astropy version.  I have 1.0.1.

rmjarvis commented:
cf. https://github.com/astropy/astropy/issues/1963

I think the warning should go away if you upgrade to a newer astropy.

rmandelb commented:
Done.  I now have 1.0.7, and the warning is gone.  Weirdly enough, I can no longer import astropy in IPython (if it's in IPython then it looks for IPython.version, which my version of IPython doesn't have; I guess go update that too but I'm lazy).  Works fine in python though.

rmjarvis commented:
I've made some further edits on this branch since my initial PR.  Mostly code simplifications and bug fixes related to things I discovered working on issue "#654" where I'm putting in some more complicated config examples using the DES modules.  

The change that is most worth calling out explicitly is in commit 5cd96da.  I moved a few of the fields that had been in `image` to `stamp`.  For backwards compatibility, it will still look in image too, but `draw_method`, `retry_failures` and `gsparams` are now moved to the `stamp` field in demos 7, 8, and 11.

rmandelb commented:
Question about this:

> Added ability to easily split the total work into several jobs with galsim -n njobs -j jobnum. This will 
> just do a single job numbered jobnum out of a total of njobs jobs. This makes it much easier to split 
> the work of a single config file over several machines.

Let's say I want to make simulations on a cluster with multiple nodes.  Each node has 16 cores that can be used to split up the postage stamp rendering within a single large image.  So let's say we want to use 10 nodes and make 200 images, dividing these up into 20 images per node (and then do the postage stamp-level sub processing automatically using all cores on the node).  Is this now feasible using this functionality?  I would make n versions of the same yaml file and just change the value of jobnum?

rmjarvis commented:
> I would make n versions of the same yaml file and just change the value of jobnum?

You only need the one yaml file.  You would change jobnum on the command line.  So the 10 jobs that you would run on the 10 different machines are:
```
galsim -n 10 -j 1 config.yaml
galsim -n 10 -j 2 config.yaml
galsim -n 10 -j 3 config.yaml
...
galsim -n 10 -j 10 config.yaml
```
All using the same yaml file.

rmjarvis commented:
BTW, I haven't really benchmarked whether it is better to run one job per node, letting it use all the cores with the GalSim multiprocessing, or if it is better to turn off multiprocessing and just run 200 jobs each to build one file.  So far, I've done the latter, letting the cluster queue basically do my multiprocessing for me, but I should really compare the two to see which ends up being faster.

rmandelb commented:
> You only need the one yaml file. 

OK, I'm glad I misunderstood, that's better.

> BTW, I haven't really benchmarked whether it is better to run one job per node, letting it use all the 
> cores with the GalSim multiprocessing, or if it is better to turn off multiprocessing and just run 200 
> jobs each to build one file.

I might test this out in a few weeks, once we are running some sims for HSC.  If I do I will let you know :)

rmandelb commented on CHANGELOG.md:
```diff
@@ -12,10 +12,51 @@ Bug Fixes
 New Features
-------------
-
+<<<<<<< HEAD
```
merge failure?

rmandelb commented on CHANGELOG.md:
```diff
@@ -12,10 +12,51 @@ Bug Fixes
-------------
-
+<<<<<<< HEAD
+- Added OutputCatalog class (#301, #691)
```
say (briefly) what this is, for those who might find it useful but weren't necessarily looking at those issues?

rmjarvis replied:
Done. ad235ec65842dc562ded37c6b46d1588da9834f8

rmandelb commented on examples/demo10.py:
```diff
@@ -166,14 +171,21 @@ def main(argv):
-    # Now have the PowerSpectrum object build a grid of shear values for us to use.
-    grid_g1, grid_g2 = ps.buildGrid(grid_spacing=stamp_size*pixel_scale, ngrid=n_tiles,
-                                    rng=rng)
+    # Initialize the OutputCatalog for the truth values
```
Should the docstring for the demo mention that this demo shows the use of the OutputCatalog functionality?

rmjarvis replied:
Done. f66ab9146f0ed6c1c913cef15e97af2040115cf1

rmandelb commented on examples/demo10.yaml:
```diff
@@ -216,3 +218,77 @@ output :
+            cosmos_index :
+                type : Eval
+                str : 'gal.index'
+                # There is a special Eval code we haven't mentioned yet.  If you name precede
```
Should this be "If you precede"?

rmandelb commented on examples/demo11.yaml:
```diff
@@ -25,17 +25,25 @@
+# We've also seen (in demo4.yaml) that YAML lets you tag values with an & in one place and
+# use them later with *.  It can be convenient to combine this notation with eval_variables
+# to let you define all the useful variables near the top of the config file and use them
+# either in Eval strings or as YAML tags.  Here, the only one use use both ways is pixel_scale,
```
"use use" -> "we use"?

rmandelb commented on examples/great3/cgc.yaml:
```diff
@@ -0,0 +1,251 @@
+# The Great3 control-ground-constant branch configuration file
```
should have license info in all the files in this dir?

rmjarvis replied:
Done. 1a5ccb6b1e5d1508c4248b731a5162242ed7c4e5

rmandelb commented on galsim/catalog.py:
```diff
@@ -343,3 +349,251 @@ def __hash__(self): return hash(repr(self))
+    --------------
+
+    @param names    A list of names for the output columns.
+    @param types    A list of types for the output columns. [default: None, which assumes all
```
Perhaps the comment below about lack of type-checking should be mentioned here?

rmjarvis replied:
Done. b91edbf6035635da4c71514d9969632d40c5f18c

rmandelb commented on galsim/config/extra_psf.py:
```diff
@@ -17,62 +17,100 @@
-        gal_shift = galsim.config.GetCurrentValue(config['gal'],'shift')
-        if False:
-            logger.debug('obj %d: psf shift (1): %s',config['obj_num'],str(gal_shift))
+    # Special: if the galaxy was shifted, then also shift the psf
```
Wait... really?  Why?  I think I am missing something.

rmjarvis replied:
This was existing functionality that I preserved.  It's only relevant for demo5 I think, where we draw the psfs in the output image at the same (shifted) position as the galaxies.  However, I'd be happy to revisit this, since it's not necessarily what the user would probably expect.

Maybe have a special option for output.psf.shift = 'galaxy' or 'match' meaning to match the galaxy's shift, but otherwise allow any normal shift instead.  So demo5.yaml's output field would look like:
```
output :
    ...

    psf :
        file_name : g08_psf.fits
        shift : galaxy
```
Currently, `offset` works the same way.  If there is an `offset` parameter in config['image'] or config['stamp'], it applies here as well.  However, this doesn't actually show up in any of the demos currently.  We could make that work the same way with an `offset : galaxy` option, or allow a different offset value if the user specifies something else.

And if neither `offset` nor `shift` are in the `output.psf` field, then just leave the PSF centered.

That seems like a sensible behavior.  Do you agree?  It's a non-backward compatible API change, but I'm ok with that.

rmandelb replied:
> However, I'd be happy to revisit this, since it's not necessarily what the user would probably expect.

Yes, that's exactly what I was worried about.  I like your proposed solution.

rmjarvis replied:
Done.  655428d

I kept demo5 with the same functionality, so it uses the new psf.shift = galaxy option.  

It turns out this change also affected demo10, which has a psf output and a galaxy shift.  So in that case, I took the opportunity to switch to an explicit offset parameter rather than copying the galaxy's shift, which both seems pedagogically useful and covers the missing functionality test.

rmandelb commented on galsim/config/input_fitsheader.py:
```diff
@@ -15,7 +15,16 @@
 import galsim
+import logging
+
+# This file adds input type fits_header and value types NFWHaloShear and NFWHaloMagnification.
```
Copy/paste error?  Should this be FitsHeader?

rmjarvis replied:
Done.  Along with other typos.  447fb0020f1ea6f51b906f39124d87836c2a4c32

rmandelb commented on galsim/config/input_powerspectrum.py:
```diff
@@ -15,15 +15,44 @@
 import galsim
+import logging
+
+# This file adds input type nfw_halo and value types NFWHaloShear and NFWHaloMagnification.
```
Copy/paste error?

rmandelb commented on galsim/config/process.py:
```diff
@@ -20,8 +20,20 @@
+# Python 2.6 doesn't include OrderdDict natively.  There is a package ordereddict that you
+# can pip install.  But if the user hasn't done that, we'll just read into a regular dict.
+# The only feature that requires the OrderedDict is the truth catalog output.  With a regular
+# dict the columns will appear in arbitrary order.
```
Should galsim emit some kind of warning in the case that this is going to happen (i.e., if we use a regular dict instead of an ordered dict)?

rmjarvis replied:
Not here I think, since it would usually be irrelevant (if the user isn't outputting a truth catalog), but I added one where the column names are processed in extra_truth.py.  7803620a7b1581a143f2ddf01f81f54308a19f88

rmandelb commented on galsim/config/value.py:
```diff
@@ -618,51 +708,74 @@ def _GenerateFromSum(config, base, value_type):
+       objects constructed using this value are safe to keep or if they have to be rebuilt.
+
+    The allowed types to include in valid_types are: float, int, bool, str, galsim.Angle,
+    galsim.Shear, galsim.PositionD.  In addition, including None in this list means that
```
no PositionI?

rmjarvis replied:
> no PositionI?

No.  We never needed one.  All the position items we build are floats, not integers.  So PositionD, not PositionI.  If a need arises, it wouldn't be hard to add this, but there hasn't been any need yet.

rmandelb commented on galsim/scene.py:
```diff
@@ -170,37 +180,73 @@ def __init__(self, file_name=None, image_dir=None, dir=None, preload=False, nois
+        # record array in each record (e.g. in getParametricRecord) and then doesn't 
+        # garbage collect it until the top-level FITS_Record goes out of scope.  
+        # This leads to a memory leak of order 10MB or so each time we make a parametric
+        # galaxy.  
```
Wow!

rmandelb commented:
OK, this is quite an impressive piece of work.  Sadly I feel that I'm not the most qualified version to review the updates to the config backend.  Not sure what to do about that.

It might be worth mentioning the GREAT3-like config scripts in the README where we talk about demos?

TallJimbo commented:
> Sadly I feel that I'm not the most qualified version to review the updates to the config backend. Not sure what to do about that.

I've actually been meaning to do my own code review on this, too, but I haven't been able to get to it yet.  Is mid-next-week too late?

rmjarvis commented:
> I've actually been meaning to do my own code review on this, too, but I haven't been able to get to it yet. Is mid-next-week too late?

Certainly not!  I would definitely appreciate a review of this from you.  Thanks, Jim.

And thanks Rachel for your comments.  I'll address them tonight probably.

TallJimbo commented on bin/galsim.py:
```diff
@@ -95,6 +95,17 @@ def parse_args():
+            'Used in conjunction with -j (--job)')
+        parser.add_argument(
+            '-j', '--job', type=int, action='store', default=1,
+            help='set the job number for this particular run. Must be in (1 .. njobs). ' +
```
This is really trivial, but I'd recommend `[1, njobs]` for the range notation.  The use of regular parenthesis to mean a closed interval (even though it's obvious what you mean with a little thought) threw me off for just a second.

rmjarvis replied:
Done.  6a10783

TallJimbo commented on examples/demo10.py:
```diff
@@ -201,10 +217,11 @@ def main(argv):
 
         # The index needs to increment every 20 objects so we use k/20 using integer math.
-        index = k / 20
+        index = k // 20
```
Since you've switched here (and elsewhere in the file, I assume), maybe do `from __future__ import division` at the top of the file?

rmjarvis replied:
This was a bit of pre-planning for making GalSim compatible with [Python 3](https://github.com/GalSim-developers/GalSim/issues/534).  The `//` operator has the same meaning in both Python 2 and 3.  But `/` does not.

Ideally, I'd like to never use `from __future__ import division` if possible, since I don't actually like having integer division return a float as it does in Python 3.  That runs against all my instincts as a C/C++ programmer.  `5 / 3` is 1, not 1.66666667.   But writing `//` at least makes me more cognizant of the issue without any ambiguity.

TallJimbo commented on examples/demo10.yaml:
```diff
@@ -100,12 +103,18 @@ gal :
+                # attributes.  So the number of galaxies in the ring defined above is gal.num.
+                # You can use any value this way.  If it hasn't actually been evaluated yet,
+                # it will evaluate it as needed for the current object.
+                repeat : { type : Current, key : gal.num }
```
Does the old raw-YAML approach work too?  If so, is there an advantage to using one or the other, besides readability?

rmjarvis replied:
The YAML version works too.  I just wanted to introduce the "Current" type here for pedagogical reasons, not because the other way doesn't work any more.  

In particular, it seemed useful to have a couple Current types early in this demo before the ones in the truth catalog section.  This and pixel_scale later are both constants, but the ones in the truth output are not.  That's the main advantage of Current over the YAML `*` notation.  The former gets the current value of things that may change from one galaxy to another.  The `*` notation only works for constants within the YAML file.

I also moved the introduction of `eval_variables` to demo11, since its use is a bit richer in that demo than what we had here.  It had been a fairly lame example in this demo, imo.  But combining it with the YAML tags is a nice design pattern that I've used in practice to put all the constants you might want to modify all together near the top of the config file.

TallJimbo commented on examples/demo10.yaml:
```diff
@@ -216,3 +219,89 @@ output :
+                # There is a special Eval code we haven't mentioned yet.  If you precede a
+                # variable name with x, then that means it's an object, rather than a value.
+                # So in this case, get the index from the attribute of the object, gal.first.
+                xgal : { type : Current, key : 'gal.first' }
```
I found this part of the example a bit confusing, for two reasons:
 - The string `gal` is overloaded to mean both the original config file element and the alias to its "first" attribute.
 - "an object, rather than a value" didn't fully make sense to me at first, because it may "object" sound like more than just a different allowed type for variables.  Maybe it'd be better to include "x = object" in the original list of allowed variable types in the comment for `psf.ellip.e`, and just refer to this as a variable with `object` type here.

TallJimbo replied:
Ah, I see now that an `object` type variable is in fact very different from the other types.  Another sentence of explanation might be helpful.

rmjarvis replied:
Good point on the overloaded name here.  Maybe calling the variable `first_gal` would be clearer?

And yes, the `x` code is a bit different from the rest, since it refers to a GSObject, which is what I meant by "object" here.  I'll add a bit more to try to make that clearer.

rmjarvis replied:
Fixed.  08d7db7c322c006918b5a0b0088af7aa93c1d3ea

TallJimbo commented on examples/demo8.py:
```diff
@@ -144,28 +141,17 @@ def main(argv):
-    #     galsim.config.BuildMultiFits(file_name, nimages, config) -- build a multi-extension fits 
-    #                                                                 file
-    #     galsim.config.BuildDataCube(file_name, nimages, config)  -- build a fits data cube
+    #     galsim.config.BuildFile(config, file_num)
```
Now that there's just single element here, it's probably best to rephrase this sentence.

TallJimbo commented on examples/demo8.py:
```diff
@@ -144,28 +141,17 @@ def main(argv):
-    # Each of these functions call the following function to process the image field
-    # and actually build the images:
+    # This functions then call the following function to process the image field and actually
+    # build each image:
```
singular/plural mismatch.

rmjarvis replied:
Done. 4b3fc25ab2e224c270edae4726329c26d07526e2

TallJimbo commented on examples/great3/cgc.yaml:
```diff
@@ -0,0 +1,268 @@
+        # The aberrations are taken to have an rms of 0.41 waves
+        # Great3 put a lot more weight into the defocus than the others.
+        # Unfortunately, this makes a couple PSFs into donuts, which was a problem.
+        # According to Aaron Roodma, for DES the appropriate rms values are closer to
```
"Roodma[n]"

rmjarvis replied:
Done.  b1730c47d7c2e4693516dc10139e6082314bb2e0

TallJimbo commented on examples/great3/rgc.yaml:
```diff
@@ -0,0 +1,40 @@
+# Now we can change the things we need to change:
+
+# 1. Use real galaxies rather than parametric
+gal.first.gal_type : real
```
Just curious: this kind of override notation is something you added on top of YAML, not something a YAML parser understands intrinsically, right?

rmjarvis replied:
Yes.  The YAML parser gets you to `config['gal.first.gal_type'] = 'real'`.  Then I have to parse it to convert it to `config['gal']['first']['gal_type'] = 'real'`.

TallJimbo commented on galsim/catalog.py:
```diff
@@ -343,3 +349,255 @@ def __hash__(self): return hash(repr(self))
+    def getNObjects(self): return self.nobjects
+    def getNCols(self): return self.ncols
+
+    def add_row(self, row, sort_key=None):
```
I see a mix of camelcase and underscore+lower method names here.  Is this intentional, to be consistent with similar method names elsewhere?

rmjarvis replied:
We're not actually very consistent about this in GalSim unfortunately.  But mixing within the same class is pretty bad, I admit.  Overall, we probably have more camelCase than underscores, so I'll switch to that.

rmjarvis replied:
Done.  b018de82b96419731b5e1c0299ff312d8471a85c

TallJimbo commented on galsim/catalog.py:
```diff
@@ -343,3 +349,255 @@ def __hash__(self): return hash(repr(self))
+
+    """
+    # Watch out for this "Gotcha".  Using _rows=[] as the default argument doesn't work!
+    # https://pythonconquerstheuniverse.wordpress.com/2012/02/15/mutable-default-arguments/
```
FWIW, my preferred workaround for this is use `_rows=()` as the argument and default value, then `self.rows = list(_rows)` to set the attribute.  But whether that's better or not depends on how much you trust the user to pass a list rather than some other sequence, and given the underscore in the name, I'm guessing that's not a concern here.

rmjarvis replied:
I agree.  That is more elegant.  Done.  63a5361b80ee02fdd8c29e47348b6950bcacb304

TallJimbo commented on galsim/catalog.py:
```diff
@@ -343,3 +349,255 @@ def __hash__(self): return hash(repr(self))
+        ncols       The number of columns in the catalog.
+        names       The names of the columns.
+        types       The types of the columns.
+        cols        The columns of data that have been accumulated so far.
```
Should this be "rows"?  I don't see an attribute assignment or property that provides "cols".

rmjarvis replied:
Yes.  Fixed.  fb5c58992fc1a3c7613b49f54ee5cf375dc141e4

TallJimbo commented on galsim/catalog.py:
```diff
@@ -343,3 +349,255 @@ def __hash__(self): return hash(repr(self))
+        file_type = file_type.upper()
+        if file_type not in ['FITS', 'ASCII']:
+            raise ValueError("file_type must be either FITS or ASCII if specified.")
+        self.file_type = file_type
```
Why set an instance attribute here?  I don't see it used downstream, and the way you've just persisted it doesn't seem like it belongs as an attribute of the in-memory object.

rmjarvis replied:
That was a residue of copying aspects of the Catalog class.  Not necessary or appropriate here.  a2e3bb1a116d32ed0bb3110eff92fdb77a76e811

TallJimbo commented on galsim/config/extra.py:
```diff
@@ -0,0 +1,377 @@
+    else:
+        return []
+
+valid_extra_outputs = {}
```
Maybe move this to the top of the file?  I was a bit confused by some logic above that referenced it before I realized it was a module-level variable.

rmjarvis replied:
Done. 38582f60c782e0784fc275c906cdb89998da39a0

TallJimbo commented on galsim/config/extra.py:
```diff
@@ -0,0 +1,377 @@
+                            [default: None, in which case the output "object" will be a list
+                            of length nimages that can be used to construct what is needed.]
+    @param kwargs_func      A function to get the initialization kwargs. [default: None, which
+                            means initialize with no arguments.]
```
No call signature for `kwargs_func`?

rmjarvis replied:
Done.  5426f30b4074146c6726e82d0b18256edf47253a

TallJimbo commented on galsim/config/extra.py:
```diff
@@ -0,0 +1,377 @@
+        'final' : final_func,
+        'write' : write_func,
+        'hdu' : hdu_func
+    }
```
While this mechanism with registered free functions is more general, this seems like place where a traditional inheritance pattern would be cleaner (i.e. make all of these callbacks no-op methods on a base class, then derived classes override the ones they want to implement, and just register the whole class).

rmjarvis replied:
OK.  I switched the extra output stuff to use the inheritance model, and I agree that it is a nicer approach, so I'm convinced.  

I had to be a bit careful about some of the multiprocessing communication, but I think I got it all sorted out.  The key is to give the builder the scratch and data objects to use for writing anything that you want to persist in later stages.

I'll work on adopting this model for the other registrations.  For now if you want to check it out, here is the commit for the extra output types: f058762dab0415073b2e57dd6881b672f10b4567

TallJimbo commented on galsim/config/gsobject.py:
```diff
@@ -79,25 +61,49 @@ def BuildGSObject(config, key, base=None, gsparams={}, logger=None):
+    if ('current_val' in param and 
+            (param['current_safe'] or param['current_index']//repeat == index//repeat)):
+        # If logging, explain why we are using the current object.
+        if logger and logger.isEnabledFor(logging.DEBUG):
```
This seems like the kind of check that it'd be better to put inside some sort of proxy object, so you don't have to keep repeating it in code that uses the logger.  I gather that the Python logging framework's built-in support for filtering log messages isn't enough to avoid the performance slowdowns you mentioned in the PR overview, but is it also impossible (or not useful) to have a proxy object to do just this check?

rmjarvis replied:
I agree.  I made a LoggerWrapper that wraps the multiprocessing proxy object and does these checks before calling the proxy object's function.  So most places in the code can skip the `isEnabledFor` part of this line.  f62f5f47e1102a2180bfe902b4755c9b86371301

TallJimbo commented on galsim/config/gsobject_ring.py:
```diff
@@ -44,7 +45,7 @@ def _BuildRing(config, base, ignore, gsparams, logger):
 
     dtheta = full_rotation / num
-    if logger and logger.isEnabledFor(logging.DEBUG):
+    if False:
```
Not sure how much you care about leaving `if False:` statements in; just wanted to make sure it wasn't accidental.

rmjarvis replied:
Thanks.  This (and a few others like it) happened when I was trying to figure out why the logging commands were so slow at times.  I put them all back in now as part of this commit: f62f5f47e1102a2180bfe902b4755c9b86371301

TallJimbo commented on galsim/config/input.py:
```diff
@@ -294,6 +298,50 @@ def GetInputObj(input_type, config, base, param_name):
+                            _req, _opt, etc. kind of initialization will not work. The call
+                            signature is:
+                                kwargs, safe = GetKwargs(config, base)
+                            [default: None, which means use the regular initialization]
```
A pointer to documentation on the "regular" kind of initialization would be helpful here (maybe a signature for `init_func`?)

TallJimbo commented on galsim/config/input.py:
```diff
@@ -294,6 +298,50 @@ def GetInputObj(input_type, config, base, param_name):
+        'nobj' : has_nobj,
+        'file' : file_scope,
+        'setup' : setup_func
+    }
```
Again, I wonder if a class hierarchy would make it easier for extenders to understand and implement the interface they need to support.

rmjarvis replied:
I'm considering this idea, but I'm not completely sure yet.  It seems like if we go with the hierarchy approach for input and extra, we should do it for all of the different registrations.  That might be a win across the board, but I'm still thinking about the consequences of it.

TallJimbo commented on galsim/config/input_powerspectrum.py:
```diff
@@ -55,6 +84,23 @@ def PowerSpectrumInit(ps, config, base):
+    # being used.
+    if not isinstance(ps, galsim.PowerSpectrum):
+        # Then ps is really a proxy, which means the rng was pickled, so we need to
+        # discard the same number of random calls from the one in the config dict.
```
It seems like there out to be a more direct way to determine whether an object is a proxy.  If it's something you have control over (I haven't followed the logic carefully enough to know if it's something from the built-in `multiprocessing` module), maybe an `is_proxy` class attribute?

rmjarvis replied:
The multiprocessing module makes these proxy objects automatically for us.  I think it uses introspection to create something with all the same methods, but which pickles the args and kwargs to these methods and sends them through the multprocessing.Pipe back to the original object, which then calls the method.  Then the return value is again pickled and sent back through the pipe to the process that called it.

It's possible there is a more elegant way to check if we have a proxy object.  I'm certainly not a python multiprocessing expert, but I couldn't come up with anything more straightforward than this.  I suppose you could try to access an attribute and catch the exception that a proxy would raise, since it doesn't copy over all the attributes, just the public methods.  But that seems even more convoluted than this isinstance check.

TallJimbo replied:
Fair enough.  If you're not the one making the objects, I agree that what you have is fine.

TallJimbo commented on galsim/config/stamp.py:
```diff
@@ -404,61 +387,166 @@ def BuildStamp(config, xsize=0, ysize=0,
+    return xsize, ysize, image_pos, world_pos
+
+
+def ProfileBasic(config, psf, gsparams, logger):
```
Maybe add a verb to this method name?  Without one, it sounds like "profile" *is* the verb, and that implies something very different.

TallJimbo commented on galsim/config/image.py:
```diff
@@ -242,27 +194,67 @@ def BuildImage(config, logger=None, image_num=0, obj_num=0,
+
+    # Mark that we are no longer doing a single galaxy by deleting image_pos from config top
+    # level, so it cannot be used for things like wcs.pixelArea(image_pos).
+    if 'image_pos' in config: del config['image_pos']
```
I'm sure isn't important from a performance standpoint, but you can write this "delete if exists" operation more simply and with one less lookup as `config.pop("image_pos", None)`.

rmjarvis replied:
Done. fed2624

rmjarvis commented:
Thanks Jim.  I've finished the refactoring to use the Builder classes approach you suggested.  There are a lot of whitespace edits, since a lot of functions got pushed in one indentation level, so it killed the github diff again, since there are too many lines changed now.  It might be easier to just look at the last few commits individually.

TallJimbo commented on galsim/config/extra.py:
```diff
@@ -0,0 +1,412 @@
+    @param builder          A builder object to use for building the extra output object.
+                            It should be an instance of a subclass of ExtraOutputBuilder.
+    """
+    valid_extra_outputs[key] = builder
```
Unnecessary but neat: make this a decorator.
```
def RegisterExtraOutput(key):
    def decorate(builder):
        valid_extra_outputs[key] = builder

@RegisterExtraOutput("badpix")
class BadPixBuilder(ExtraOutputBuilder):
    ...
```

rmjarvis replied:
I think I'll pass on this recommendation.  No offense.  :)  I'm not super familiar with decorators yet, and I'm worried that this will be intimidating to other users who want to use this registration functionality, who may be even less familiar with decorator syntax.

Also, I think as written, this will assign the _class_ to the `valid_extra_outputs` dict rather than an _instance_ of that class.  For the `ExtraOutputBuilder`s, we could do it that way: have the class be the thing that is registered and make an instance later when we need it.  

But some of the other things I register use the same class with different initialization parameters for different types.  So I decided it made more sense to have everything register instances rather than classes to be consistent.  I think in the decorator syntax, this would look like:
```
def RegisterExtraOutput(key):
    def decorate(builder):
        valid_extra_outputs[key] = builder

class BadPixBuilder(ExtraOutputBuilder):
    ...

@RegisterExtraOutput("badpix")
BadPixBuilder()
```
which doesn't seem like a big improvement over the existing syntax using the function call.

TallJimbo replied:
Fair enough.  I had indeed missed the point about instances vs. classes here, and even if familiarity with decorators wasn't a concern I think that would shoot down this proposal.

TallJimbo commented:
I've now finished going through the latest versions of things, with my only comment the one above on decorators.  Looks great!

rmjarvis commented:
OK.  Thanks.  I think this is all done then, unless anyone else would like to take a look.  I'll plan to merge this tomorrow night unless I hear otherwise.
